File: golang/20_valid_parentheses.go
package main

import "fmt"

// Contains given string in array or not
func Contains(a [3]string, val string) bool {
	for _, c := range a {
		if c == val {
			return true
		}
	}
	return false
}

func isValid(s string) bool {
	var stack []string
	openingBraces := [3]string{"(", "{", "["}
	closingBraces := [3]string{")", "}", "]"}
	for _, c := range s {
		strC := string(c)
		if Contains(openingBraces, strC) {
			stack = append(stack, strC)
		} else if Contains(closingBraces, strC) {
			n := len(stack) - 1
			if n == -1 {
				return false
			}
			topCh := stack[n]
			if (topCh == "(" && strC == ")") || (topCh == "{" && strC == "}") || (topCh == "[" && strC == "]") {
				stack = stack[:n]
			} else {
				return false
			}
		}
	}
	return len(stack) == 0
}

func main() {
	var inp string
	fmt.Scanln(&inp)
	fmt.Println(isValid(inp))
}

File: python/20_valid_parentheses.py
class Stack:
    def __init__(self):
        self.stack = []
    
    def isEmpty(self):
        return self.size() == 0
    
    def push(self, item):
        self.stack.append(item)
    
    def pop(self):
        return self.stack.pop()
    
    def peek(self):
        return self.stack[self.size() - 1]
    
    def size(self):
        return len(self.stack)

class Solution:
    def isValid(self, s: str) -> bool:
        st = Stack()
        opening_braces = ['(', '{', '[']
        closing_braces = [')', '}', ']']
        for ch in s:
            if ch in opening_braces:
                st.push(ch)
            elif ch in closing_braces:
                if st.isEmpty():
                    return False
                top_ch = st.peek()
                if (top_ch == '(' and ch == ')') or (top_ch == '{' and ch == '}') or (top_ch == '[' and ch == ']'):
                    st.pop()
                else:
                    return False
        return st.isEmpty()

sol = Solution()
inp = input()
print(sol.isValid(inp))